### PR TITLE
fix for inheritence with type intersection, basic type lookup support…

### PR DESCRIPTION
**Pull request by pshemass**
First of all I want to say that the model that was build here to represent Typescript and operations on it like Rewrites, FollowAliases are really powerful. 

This PR is result of of my investigation why `Button` and `IconButton` are not parsed correctly in 4.x of `material-ui`. 
It seems that the problem is with inheritance with intersection type. 

```
export type ExtendBase<M extends OverridableTypeMap> = ((
  props: { href: string } & OverrideProps<ExtendButtonBaseTypeMap<M>, 'a'> // <-- here we have type alias to intersection
) => JSX.Element) &
  OverridableComponent;

declare const Button: ExtendBase<ButtonTypeMap>;
```

another problem that I found was support for field type lookup that is pretty difficult to express in scala. I added some trick to resolve issue with missing props in Slinky components but I don't know if that is good enough.

```
export interface ExtendButtonBaseTypeMap<M extends OverridableTypeMap> {
  props: M['props'] ; // type of field with name "props" in type M
  defaultComponent: M['defaultComponent'];
}
```

### CODE REVIEW

**Comment by oyvindberg** [comment]
Thank you, I'm intrigued to see what you've been up to! :)

I'm also a bit more busy than normal, so I'll need a day or two to respond properly

**Comment by oyvindberg** [comment]
Ok, I did warn you that this is pretty complicated territory, and I'm both horrified and happy that you're looking into it! :)

TLDR: You need to look closer at [ExpandTypeMappings](https://github.com/ScalablyTyped/Converter/blob/master/ts/src/main/scala/org/scalablytyped/converter/internal/ts/transforms/ExpandTypeMappings.scala)

The architecture of the converter is that of a long pipeline which gradually goes from Typescript code to Scala code, through a long list of transformations, as you've seen. It might be hard to see the forest for the trees here, but the goal is that each transformation will "simplify" the tree, in the very concrete sense that we take one Typescript construct and rewrite it into an (hopefully!) equivalent tree which no longer uses that construct.

For instance:
- DefaultedTypeArguments - After running this we can safely assume that all type parameters are now explicitly specified.
- DropProperties - After running this we can safely assume that no interfaces have a `prototype` member

...and so on. I've tried to decompose the problem of rewriting Typescript into a bunch of transformations like this.

As long as the trees are fundamentally Typescript they are transformed within the `ts` module, and that's where Typescript features are rewritten. After the trees pass through `ImportTree` they are meant to be semantically Scala trees, and the Scala transformations mostly deal with making it compile. For instance they deal with overrides, overloads, erasure and so on. It's not 100% clear cut, but that's the idea.

---- 

Which brings us to `ExpandTypeMappings`. The job of this transformation is the largest and most complicated, and relatively speaking the least well developed in the code base. Its job is to take any kind of type description which uses [mapped types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types) and rewrite it into a completely normal interface. It's a huge thing, because along with conditional types, lookup types, literal types, and the newly introduced template literal types it forms a whole new programming language at the type level in typescript. I'm not kidding, somebody just implemented an [SQL engine](https://twitter.com/c_pick/status/1307433762914009090) in this language.

So needless to say we don't support anything close to that, and never will. The reason why we never will is that we can only possibly do this for things which are static. We'll never have a working translation of this:
```typescript
type MyPartial<T> = Partial<T>
// of course referring to the normal `Partial`
//type Partial<T> = {
//  [P in keyof T]?: T[P];
//};
```
But we can have a translation of this:
```typescript
type A = Partial<{a: number}>
```

So a few more observations about `ExpandTypeMappings`:
- It currently doesn't support `infer`
- It currently doesn't really support conditional types. Crucially for implementing that we would need an implementation of `extends` which would follow Typescripts rules for deciding whether a given pair of types have a subtyping relationship
- it's prone to infinite recursion, out of memory errors, and all that fun. It has a few pretty crude mechanisms for avoiding that
- As a result it only successfully converts "basic" uses of the mapped types feature, but for that it works pretty well with not too much overhead. The happy news is that most usage is static in the end, and most usage is "basic".  



Fine, where does that leave us with material-ui 4? It's not static like the latter case, but close - all the relevant components have a *default* underlying component - for instance `Button` has the react intrinsic `"button"`. I have studied the problem enough to be confident that we can generate what we need, doubly so if we [cheat](https://github.com/ScalablyTyped/Converter/blob/master/ts/src/main/scala/org/scalablytyped/converter/internal/ts/transforms/LibrarySpecific.scala) a bit.

It is also decidedly not "basic" anymore, so we'll need to extend `ExpandTypeMappings` with just enough power to handle it, probably with a sprinkle of cheating to save ourselves some hard implementation work.

So just to get back on track with this PR, the fundamental problem with it is that you're implementing it on the Scala side of things, when it needs to happen on the Typescript side. To complete it you would need to reimplement half of `TsType` in `TypeRef`, and reimplement `ExpandTypeMappings` with friends. The most visible symptom of this is that `typingsSlinky.inheritanceWithIntersection.components.Button` doesn't have members from `React.ButtonHTMLAttributes`, and it'll be hard to put them there.

I pushed my (very WIP!) commit from last time I was experimenting with this here https://github.com/ScalablyTyped/Converter/commits/mui4 . Feel more than free to pick it up. I'm also open to a lot of improvements on this front, including writing a new `ExpandTypeMappings` implementation, but it'll have to happen within the Typescript side.

Ok, that was a wall of text, I hope it helps. Just reach out for instance on gitter if you want to talk more about this

**Comment by oyvindberg** [comment]
Also I'll look into the type intersection thing tomorrow. The reason why it wasn't supported in the code path originally is that this isn't (perhaps surprisingly) valid scala `class C extends (A with B)`.

**Comment by pshemass** [comment]
@oyvindberg now we are talking :) It makes perfect sense to me to simplify this on TS level. Let me try it.

**Comment by pshemass** [comment]
@oyvindberg I cannot figure out where is the place that `Button` is connected to `React.ButtonHTMLAttributes`. 

Is this is because of this ?
```
export type ButtonTypeMap<
  P = {},
  D extends React.ElementType = 'button' // <-- is this is it
>
...
defaultComponent: D;
...
```
